[bug 1170333] es

willkg opened the pull request:
r?

willkg commented:
This switches us from ElasticUtils to elasticsearch-dsl and also is a complete overhaul of the Elasticsearch using code.

Changes of note:

1. we switched from facets to aggregations (facets are deprecated and will go away soon)
2. we switched from the MoreLikeThis to the morelikethis search query (MoreLikeThis is deprecated and will go away soon)
3. I moved some code around to make the API a little easier to work with
4. I took the time to move the graph that's on the front page dashboard and only shows up if you're an analyzer to the analyzer search.

Things that probably aren't of interest:

1. Reindexing from the command line takes about the same amount of time as it did before. I spent some time tinkering with the code and I'm pretty sure at this point that it's bound on how fast we can get data out of the db. We could speed it up a bit more by getting dicts out instead of model instances, but that requires us to duplicate some code here and there and will make maintenance harder and mistakes more likely.
2. elasticsearch-dsl won't store empty string values (https://github.com/elastic/elasticsearch-dsl-py/issues/202). I ended up writing a `.to_index()` on `ResponseDocType` which both reduces the number of objects created thus speeding indexing up significantly and also gets around the empty string value problem.

To review:

1. read my blog post for some context on the changes: http://bluesock.org/~willkg/blog/mozilla/ditching_elasticutils.html
2. read through the code--half of it is a "this to that" translation and the other half is more involved because I rewrote a bunch of code
3. run the tests
4. run the various `es_*` commands on the command line

willkg commented on fjord/analytics/analyzer_views.py:
```diff
@@ -166,12 +167,12 @@ def _analytics_search_export(request, opinions_s):
     total_opinions = opinions_s.count()
 
-    opinions_s = opinions_s.values_list('id')[:MAX_OPINIONS]
+    opinions_s = opinions_s.fields('id')[:MAX_OPINIONS].execute()
```
The ElasticUtils `values_list` and the elasticsearch-dsl `fields` essentially do the same thing.

With elasticsearch-dsl, you have to explicitly call `execute` to run the search and get back a `Results` instance.

willkg commented on fjord/analytics/analyzer_views.py:
```diff
@@ -223,27 +220,32 @@ def analytics_search(request):
+    search_has_email = request.GET.get('has_email', None)
+    search_country = request.GET.get('country', None)
+    search_domain = request.GET.get('domain', None)
+    search_api = smart_int(request.GET.get('api', None), fallback=None)
```
I moved these down here because this view is almost a copy of the non-analyzer frontpage dashboard. In moving these to this section, it makes it easier to compare the dashboard view with this view to see differences.

willkg commented on fjord/analytics/analyzer_views.py:
```diff
@@ -251,57 +253,61 @@ def unknown_to_empty(text):
 
     if search_date_end is None:
-        search_date_end = datetime.now()
+        search_date_end = date.today()
```
This was a bug. It should never have been a datetime since we're only looking at dates and not time-of-days.

mythmon replied:
Switching this to dates instead of datetimes makes the code below a lot clearer. :+1:.

willkg commented on fjord/analytics/analyzer_views.py:
```diff
@@ -251,57 +253,61 @@ def unknown_to_empty(text):
     # Note 'less than', not 'less than or equal', because of the added
     # day above.
-    f &= F(created__lt=end)
+    f &= F('range', created={'lte': search_date_end})
```
I switched this to an lte after fixing the datetime goof above.

willkg commented on fjord/analytics/analyzer_views.py:
```diff
@@ -315,23 +321,25 @@ def unknown_to_empty(text):
     if output_format == 'csv':
         return _analytics_search_export(request, search)
 
+    original_search = search._clone()
```
I needed to do two sets of aggregation and thought it was easier to do if I cloned it because one set involves an additional filter or something like that. I need to revisit this.

willkg commented on fjord/analytics/templates/analytics/analyzer/search.html:
```diff
@@ -145,6 +145,12 @@
+      <div class="histogram" data-histogram="{{ histogram|to_json }}"></div>
+      <div class="legend"></div>
+      <div class="note">{{ _('Note: All dates and times are in Pacific Time.') }}</div>
+    </div>
```
The frontpage dashboard used to have a graph on it. A while back, I changed the graph so it only shows up for people who are in the analyzers group.

The use case is that Matt and crew want to be able to do a search and see the graph as a first pass for figuring out whether something is a big deal and needs looking into or not. This helps them field requests from other people about the data.

I left it on the front page because I didn't want to spend the time to move it. In redoing all the Elasticsearch stuff, I decided to finally move it.

willkg commented on fjord/analytics/templates/analytics/dashboard.html:
```diff
@@ -128,15 +128,6 @@
-      <div class="note">{{ _('Note: All dates and times are in Pacific Time.') }}</div>
-    </div>
-  {% endif %}
-
```
I moved the graph from the front page dashboard to the analyzer search.

willkg commented on fjord/analytics/tests/test_analyzer_views.py:
```diff
@@ -89,6 +89,35 @@ def setUp(self):
+        # not the end point we want to test against.
+        while d < end:
+            assert d.day in days, 'Day %s has no data.' % d.day
+            d += timedelta(days=1)
```
Since I moved the graph from the dashboard to the analyzer search, I moved some tests around, too.

willkg commented on fjord/analytics/tests/test_views.py:
```diff
@@ -245,25 +245,6 @@ def test_search_format_atom(self):
-    #     # both attributes.
-    #     assert 'rel="related"' in r.content
-    #     assert 'href="http://example.com"' in r.content
-
```
I deleted this dead code.

willkg commented on fjord/analytics/tests/test_views.py:
```diff
@@ -329,36 +310,6 @@ def test_frontpage_index_missing(self):
-        while d < end:
-            assert d.day in days, 'Day %s has no data.' % d.day
-            d += timedelta(days=1)
-
```
This was moved to the analyzer view tests.

willkg commented on fjord/feedback/models.py:
```diff
@@ -408,8 +403,103 @@ def infer_platform(cls, product, browser):
 
-@register_mapping_type
-class ResponseMappingType(FjordMappingType, Indexable):
+class ResponseDocTypeManager(FjordDocTypeManager):
```
I decided to model this similarly to the Django Model/ModelManager. I'm not sure if that was helpful or just added unneeded complexity.

willkg commented on fjord/search/apps.py:
```diff
@@ -0,0 +1,19 @@
+        urls = getattr(settings, 'ES_URLS', DEFAULT_URLS)
+        timeout = getattr(settings, 'ES_TIMEOUT', DEFAULT_TIMEOUT)
+
+        connections.configure(default={'hosts': urls, 'timeout': timeout})
```
elasticsearch-dsl maintains connections. So here, we configure all the connections (in our case, it's just "default") so that we can use them elsewhere.

willkg commented on fjord/search/index.py:
```diff
@@ -18,7 +25,79 @@
+    :arg alias: the alias in ES_URLS for this Elasticsearch connection
+
+    """
+    return connections.get_connection(alias=alias)
```
I kept `get_es` because it's shorter than writing out this line everywhere.

willkg commented on fjord/search/index.py:
```diff
@@ -41,34 +120,23 @@ def es_analyze(text, analyzer=None):
 
-def get_index():
-    """Returns the index we're using."""
+def get_index_name():
```
I renamed this because elasticsearch-dsl has an Index class and I decided it was worth clarifying this function.

willkg commented:
I went through, added a bunch of comments to the PR to make it clearer why things happened and then also fixed some minor things.

mythmon commented on fjord/analytics/analyzer_views.py:
```diff
@@ -251,57 +253,59 @@ def unknown_to_empty(text):
         search_date_start = search_date_end - timedelta(days=7)
 
+    # If the start and end dates are inverted, switch them into proper
+    # chronoligcal order
```
Typo: chronoligcal -> chronological

mythmon commented on fjord/analytics/analyzer_views.py:
```diff
@@ -251,57 +253,59 @@ def unknown_to_empty(text):
+    # If the start and end dates are inverted, switch them into proper
+    # chronoligcal order
+    search_date_start, search_date_end = sorted(
+        [search_date_start, search_date_end])
```
Nice! I wouldn't have thought to use sort for that.

mythmon commented on fjord/analytics/views.py:
```diff
@@ -235,16 +227,16 @@ def unknown_to_empty(text):
+            f &= F('term', version=unknown_to_empty(search_version))
             current_search['version'] = search_version
     else:
-        f &= F(product__in=visible_products)
```
Is this right? Before you were doing a `__in` query, but now you are doing a normal terms query.

willkg replied:
In Elasticsearch 0.90, the "terms" query was aliased as "in":

https://www.elastic.co/guide/en/elasticsearch/reference/0.90/query-dsl-terms-query.html

I can't tell when they deprecated that alias, but it's not in the new Elasticsearch docs, so I switched it.

willkg commented:
^^^ Fix typo.

willkg commented:
@mythmon What do you think here? Is it good enough to land?

mythmon commented:
I don't really understand the reason to split DocTypes and managers, but I don't really understand the reason behind Django's split either, so that isn't a huge deal. The interface makes sense and looks easy to use on both sides, so :+1:.

Should you add elasticsearch-dsl-py to one of the requirements files? I tried to install it from there, but I didn't find it. I'm also having some trouble with git submodule that I think is something wrong locally. I need to check that out. Either way, I'm not able to actually test this locally right now. I can work on that more this afternoon.

I don't see anything else in the code I want to comment on. The tests and running the code should cover any changes in behavior, so I'm pretty confident in that (assuming I can get it running). The comments and the blog post were very helpful, thanks!

willkg commented:
I think the split between Model and ModelManager is that ModelManager is an instance of a class that lets you operate on groups of objects whereas Models are about specific things. It lets you do things like have multiple ModelManagers available that think about different groups of objects and in doing that, you get more readable code.

The split here between DocType and DocTypeManager is probably overengineering at this point, but it seemed easier to implement it now this way and see how it feels than it did to just implement DocType and have to potentially fix it later.

I'll look at the requirements file--I probably forgot to do that.

I'm not sure why you're having problems with vendor/, though. Are there error messages?

willkg commented:
^^^ Rebased against master and also add the requirements.

willkg commented:
Travis is green. I've addressed all the issues. I did another round of testing. I'm going to call it "good enough" and land it. I'll test extensively on -stage next to make sure I didn't break anything. Then push it to prod either tonight or tomorrow.

willkg commented:
@mythmon Thank you!
